src/mgr/ActionParams.cc

```diff
@@ -38,9 +40,9 @@ Mgr::ActionParams::ActionParams(const Ipc::TypedMsgHdr &msg)
 void
 Mgr::ActionParams::pack(Ipc::TypedMsgHdr &msg) const
 {
-    msg.putString(httpUri);
-    auto foo = SBufToString(httpMethod.image());
-    msg.putString(foo);
+    msg.putString(httpMethod.image());
+    // XXX: TypedMsgHdr::maxSize < 8KB URL required minimum capacity
+    msg.putString(httpUri.absolute());
```

Comment:
Assumption of "no compelling reason" is wrong. This simple switch of order allows simpler validation and conversion to binary at the receiver.

IMO the proposed XXX wording is less clear. The problem is very specifically the 8KB mandatory requirement of the specifications.

Reply:
> > Assumption of "no compelling reason" is wrong. This simple switch of order allows simpler validation and conversion to binary at the receiver.
> 
> I do not see any significant improvement in "validation and conversion to binary at the receiver",

Validating the URL/URI with `AnyP::Uri` requires knowing the method. So having it available at the time of `getString(url)` allows immediate validation of the URL form.

> so I still cannot find any compelling reason to break compatibility, especially in a PR that is focused on String removal.
> 

Er. This is an **undocumented** binary protocol between two fork()'ed binaries. This class is literally talking to another instance of itself. There is no "compatibility" existing to be broken.


> > IMO the proposed XXX wording is less clear. The problem is very specifically the 8KB mandatory requirement of the specifications.
> 
> This code is broken regardless of the "8K minimum". For example, if that specified minimum would have been "8 bytes", this code would still be wrong, for exactly same reason!

That is not true. 4096 is greater than 8. It might be wrong for __different__ reasons than the one I am documenting.

To be clear. What I am documenting is that: 
  Per HTTP specification a valid URL of 8000 octets SHOULD NOT trigger any errors in any HTTP agent.
  Yet  `TypedMsgHdr::maxSize = 4096` definition causes exception handling to produce errors on 4097 octet URLs.


> My XXX variation explicitly documents why this code is wrong. PR XXX variation can be easily misinterpreted as falsely implying that making TypedMsgHdr::maxSize, say, 9KB will fix this problem. It will not.

Per HTTP specification a valid URL of 8001 or more octets may trigger errors in any HTTP agent.

Reply:
> > > Assumption of "no compelling reason" is wrong. This simple switch of order allows simpler validation and conversion to binary at the receiver.
> > 
> > 
> > I do not see any significant improvement in "validation and conversion to binary at the receiver",
 
> Validating the URL/URI with `AnyP::Uri` requires knowing the method. So having it available at the time of `getString(url)` allows immediate validation of the URL form.

```C++
// old/official order
const auto uri = msg.getString();
httpMethod = HttpRequestMethod(msg.getString());
httpUri.parse(httpMethod, uri);
```

```C++
// new/PR order
httpMethod = HttpRequestMethod(msg.getString());
const auto uri = msg.getString();
httpUri.parse(httpMethod, uri);
```

I see no difference as far as "immediate validation" is concerned.

> > so I still cannot find any compelling reason to break compatibility, especially in a PR that is focused on String removal.

> Er. This is an **undocumented** binary protocol between two fork()'ed binaries. This class is literally talking to another instance of itself. There is no "compatibility" existing to be broken.

The "talking to itself" argument ignores use cases where admins write custom Squid interrogation tools or  upgrade/downgrade individual Squid kid processes to minimize user disruptions. Those (real) use cases do _not_ imply that we can never change IPC message format, of course. They only imply that we should make changes when there is a compelling reason to do that (and the commit should call out these changes if its primary scope does not make them obvious). I still do not see such a compelling reason here.

 
> > > IMO the proposed XXX wording is less clear. The problem is very specifically the 8KB mandatory requirement of the specifications.

> > This code is broken regardless of the "8K minimum". For example, if that specified minimum would have been "8 bytes", this code would still be wrong, for exactly same reason!
 
> It might be wrong for **different** reasons than the one I am documenting.

> To be clear. What I am documenting is that: Per HTTP specification a valid URL of 8000 octets SHOULD NOT trigger any errors in any HTTP agent. Yet `TypedMsgHdr::maxSize = 4096` definition causes exception handling to produce errors on 4097 octet URLs.

My XXX version covers the problem you are documenting and more. Why argue about it?! If you think it is important to mention some RFC SHOULD, here is another version (just fill in the RFC number or adjust similarly as needed):

```suggestion
    // XXX: TypedMsgHdr::maxSize is too small for some valid mgr URIs,
    // including 8KB URLs that RFC ... says SHOULD be supported.
    msg.putString(httpUri.absolute());
    msg.putString(httpMethod.image());
```